# HG changeset patch
# User Ken Brown <kbrown@cornell.edu>
# Date 1426692861 14400
#      Wed Mar 18 11:34:21 2015 -0400
# Node ID 513664dc75ee845837ac2b76898f0e9c285e670d
# Parent  0c1c2f8a54a2a829667ac96f3b33fe38a2a9a5e8
Fix dynamic loading on Cygwin

* src/genclisph.d [UNIX_CYGWIN32]: Fix syntax error in lisp.def.
* src/makemake.in (archive_cmds): Fix definition on Cygwin.
(lisp.def): Add "hostent_to_lisp" to lisp.def on Cygwin.
(clisp-link): Use absolute path to lisp.def.

diff -r 0c1c2f8a54a2 -r 513664dc75ee src/genclisph.d
--- a/src/genclisph.d	Thu Mar 05 15:04:35 2015 -0500
+++ b/src/genclisph.d	Wed Mar 18 11:34:21 2015 -0400
@@ -333,7 +333,13 @@
     def_f = fopen(argv[2],"w");
     if (def_f == NULL) { perror(argv[2]); exit(1); }
     fprintf(stderr,"writing DLL export file %s\n",argv[2]);
+    /* Having both EXPORTS and IMPORTS generates a syntax error on
+       Cygwin.  All we need is imports, for building dynamic modules. */
+   #ifdef UNIX_CYGWIN32
+    fprintf(def_f,"IMPORTS\n");
+   #else
     fprintf(def_f,"EXPORTS\nIMPORTS\n");
+   #endif
   }
  #endif
 
diff -r 0c1c2f8a54a2 -r 513664dc75ee src/makemake.in
--- a/src/makemake.in	Thu Mar 05 15:04:35 2015 -0500
+++ b/src/makemake.in	Wed Mar 18 11:34:21 2015 -0400
@@ -1371,7 +1371,11 @@
   # Support for dynamic loading.
   eval "`./libtool --tag=CC --config | grep '^pic_flag='`"
   eval XCC_PICFLAG=\"${pic_flag}\"
-  eval "`./libtool --tag=CC --config | grep '^archive_cmds='`"
+  if [ "$HSYSOS" != cygwin ]; then
+    eval "`./libtool --tag=CC --config | grep '^archive_cmds='`"
+  else
+    archive_cmds='$CC -shared $libobjs $deplibs $compiler_flags -o $output_objdir/$soname'
+  fi
   # http://article.gmane.org/gmane.comp.shells.bash.bugs:17158
   CC_set=${CC+set}
   CC_save=$CC
@@ -3081,14 +3085,20 @@
   echol
 fi
 
-
 echol "clisp.h clisp-test.c${CLISP_DEF} : genclisph.o config.h ${PARAMS_H}"
 echotab_to_HEXE "\$(CC) \$(CFLAGS) \$(CLFLAGS)" "genclisph.o" "genclisph"
 echotab "(echo '#ifndef _CLISP_H' ; echo '#define _CLISP_H' ; echo; echo '/* ==== config.h ==== */' ; cat config.h ) > clisp.h"
 for parf in ${PARAMS_H}; do
   echotab "(echo; echo '/* '${parf}' */' ; grep '^#' ${parf} ) >> clisp.h"
 done
-echotab "(echo; echo '/* genclisph */' ; ${HERE}genclisph clisp-test.c ${CLISP_DEF}; echo ; echo '#endif /* _CLISP_H */') >> clisp.h"
+# clx depends on the symbol "hostent_to_lisp", which is defined in the
+# syscalls module and therefore doesn't get put into lisp.def on
+# Cygwin.  So we just force it.
+if [ "${HSYSOS}" != cygwin ]; then
+  echotab "(echo; echo '/* genclisph */' ; ${HERE}genclisph clisp-test.c ${CLISP_DEF}; echo ; echo '#endif /* _CLISP_H */') >> clisp.h"
+else
+    echotab "(echo; echo '/* genclisph */' ; ${HERE}genclisph clisp-test.c ${CLISP_DEF}; echo -e \"\tlisp.exe.hostent_to_lisp\" >> ${CLISP_DEF} ; echo ; echo '#endif /* _CLISP_H */') >> clisp.h"
+fi
 echotab_to_HEXE "\$(CC) \$(CPPFLAGS) \$(CFLAGS) \$(CLFLAGS) -DUSE_CLISP_H=1 -DCOMPILE_STANDALONE" "clisp-test.c" "clisp-test-clisp"
 echotab_to_HEXE "\$(CC) \$(CPPFLAGS) \$(CFLAGS) \$(CLFLAGS) -DUSE_CLISP_H=0 -DCOMPILE_STANDALONE" "clisp-test.c" "clisp-test-lispbibl"
 echotab "${HERE}clisp-test-clisp > clisp-test-clisp.out"
@@ -3160,8 +3170,12 @@
   echotab "sed -e 's%@with_dynamic_modules@%${with_dynamic_modules}%' \\"
   echotab "    -e 's%@createsharedlib@%$XCC_CREATESHARED_%' \\"
   echotab "    -e 's%@LEXE@%${LEXE}%' -e 's%@SHREXT@%${SHREXT}%' \\"
+  # On Cygwin, "-L$sourcedir$" is not enough for the linker to find
+  # lisp.def.  So just give the absolute path to lisp.def.
   if test -z "${CLISP_DEF}"; then
     echotab "    -e 's%@CLISP_DEF@%%' \\"
+  elif [ "${HSYSOS}" = cygwin ]; then
+    echotab "    -e 's%@CLISP_DEF@%\$\$absolute_sourcedir/lisp.def%' \\"
   else
     echotab "    -e 's%@CLISP_DEF@%-L\$\$sourcedir${CLISP_DEF}%' \\"
   fi
